### ISSUE TITLE
Decouple the MatcherPolicy from MVC and make it solely based on routing.

### ISSUE THREAD

**Issue by davidfowl** [dotnet/aspnet-api-versioning#751]
The implementation could be made more generic by using endpoint metadata directly instead of the ActionDescriptor to stash the API information. Instead of doing [SetProperty](https://github.com/microsoft/aspnet-api-versioning/blob/e16b579d240574053ebd8e7ee38c8686beaee174/src/Microsoft.AspNetCore.Mvc.Versioning/Versioning/Conventions/ActionApiVersionConventionBuilderBase.cs#L60) the data could be stored [here](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.abstractions.actiondescriptor.endpointmetadata?view=aspnetcore-5.0).

I haven't looked deeply to see what other tentacles reach into MVC.

**Comment by commonsensesoftware** [comment]
Interesting that you mention that. A few people have asked how that might be achieved.

Switching from storing things in an ActionDescriptor to EndpointMetadata, or more specifically `Endpoint.Metadata`, is feasible - with the right amount of refactoring. API Versioning was around before _Endpoint Routing_ (and helped flush out some of the design with Ryan and James). The two main reasons the information is still in a property are:

1. Historical
2. Continues to work for those still using _legacy_ routing (they're still out there)

`#2` can be addressed by the fact that you can't use both routing systems simultaneously. A _strategy_ can be used to control where the information is accessed to avoid a bunch of branching conditions. Today, this would normally be detected through `MvcOptions.EnableEndpointRouting`. If MVC were removed, then it would probably be assumed `true` and only _possibly_ switch when integrated with MVC.

I've also be considering breaking things further apart; something like _Abstractions_. This work and feature would likely align to those changes. Any thoughts or suggestions are most certainly welcome.

TL;DR

The fundamental challenge to removing MVC completely is how the _metamodel_ is currently discovered. Version discovery and collation are largely done through the **ApplicationModels** namespace. In the strictest sense, API Versioning doesn't really care about that and isn't bound to it. A fairly early design point was API version information is purely metadata. It is _provided_ by way of `IApiVersionProvider`. This is most commonly achieved by way of attributes, but there is a _conventions_ fluent API as well. The _Application Model_ afforded by MVC provides a nice way to inspect and apply API version metadata by interrogating the developer's code.

API version collation is one of the more difficult challenges. Despite what some people may think, it's achieved using the controller's name. Again - it's doesn't _have_ to, but it was a natural choice. It might seem more logical to group by route template, but this can go sideways pretty quickly. For example, `api/v{ver:apiVersion}/{id}` and `api/v{version:apiVersion}/{id:int}` are semantically equivalent (and obvious to developers reading it), but is quite difficult - as well as likely slow - to match up. There are other scenarios where the route template just doesn't give you enough information. For example, what if `value/{id}` and `value/{id}/subvalue/list` should, or shouldn't, be grouped together? It's a bit more nuanced than that, but hopefully that paints a picture.

Armed with that information, what I conceptually think could work if those other issues are addressed would be something like:

```c#
app.UseEndpoints(endpoints =>
{
	// version-neutral api
	endpoints.MapGet("/ping", c => Task.CompletedTask)
	         .IsApiVersionNeutral();

	// 1.0 endpoint
	endpoints.MapGet("/hello", c => c.Response.WriteAsync("Hello world v1!"))
	         .HasApiVersion(1, 0);

	// 2.0 endpoints (grouped for convience)
	endpoints.WithApiVersion(2, 0, api =>
	{
	      api.MapGet("/hello", c => c.Response.WriteAsync("Hello world v2!"));
	});
});
```

I suspect this approach breaks or inhibits API Explorer features for those that care about it - usually for OpenAPI documents. I believe that's already an issue and an understood decision point for choosing this type of setup.

**Comment by davidfowl** [comment]
We're interested in this so we'll investigate what it would take to make it work in the .NET 7 timeframe.

**Comment by commonsensesoftware** [comment]
@davidfowl,

I just wanted to let you know that this work has been completed (but I'm not closing the issue just yet). Due to changes in the project that have just been [announced](../discussions/807), we're still a little bit out from publishing the corresponding packages. If you're interested in taking a peek, you can see the implementation [here](../tree/main/src/AspNetCore/WebApi/src/Asp.Versioning.Http). Things are now split apart from core HTTP and MVC Core, where API Versioning is possible without any part or direct references to the MVC Core assembly. This also sets up the foundation for _Minimal API_ support.

**Comment by commonsensesoftware** [comment]
As it relates to _Minimal API_, everything will tie together like this:

```c#
builder.Services.AddApiVersioning();

var app = builder.Build();

app.DefineApi()
   .HasApiVersion( 1.0 )
   .HasApiVersion( 2.0 )
   .ReportApiVersions()
   .HasMapping( api =>
    {
        // GET /weatherforecast?api-version=1.0
        api.MapGet( "/weatherforecast", () =>
            {
                return Enumerable.Range( 1, 5 ).Select( index =>
                    new WeatherForecast
                    (
                        DateTime.Now.AddDays( index ),
                        Random.Shared.Next( -20, 55 ),
                        summaries[Random.Shared.Next( summaries.Length )]
                    ) );
            } )
           .MapToApiVersion( 1.0 );

        // GET /weatherforecast?api-version=2.0
        api.MapGet( "/weatherforecast", () =>
            {
                return Enumerable.Range( 0, summaries.Length ).Select( index =>
                    new WeatherForecast
                    (
                        DateTime.Now.AddDays( index ),
                        Random.Shared.Next( -20, 55 ),
                        summaries[Random.Shared.Next( summaries.Length )]
                    ) );
            } )
           .MapToApiVersion( 2.0 );

        // POST /weatherforecast?api-version=2.0
        api.MapPost( "/weatherforecast", ( WeatherForecast forecast ) => { } )
           .MapToApiVersion( 2.0 );

        // DELETE /weatherforecast
        api.MapDelete( "/weatherforecast", () => { } )
           .IsApiVersionNeutral();
    } );
```

**Comment by davidfowl** [comment]
@commonsensesoftware We're doing route grouping in .NET 7 https://github.com/dotnet/aspnetcore/issues/36007.

cc @halter73 @captainsafia

**Comment by commonsensesoftware** [comment]
@davidfowl,

[Preview 2](../releases/tag/v6.0.0-preview.2) is now available. The 🐀 🧶 has been eliminated. To the greatest extent possible, everything now only considers endpoints and their metadata.

New Packages:

- [Asp.Versioning.Http](https://www.nuget.org/packages/Asp.Versioning.Http) - _Minimal APIs_
- [Asp.Versioning.Mvc](https://www.nuget.org/packages/Asp.Versioning.Mvc) - MVC Core
- [Asp.Versioning.Mvc.ApiExplorer](https://www.nuget.org/packages/Asp.Versioning.Mvc.ApiExplorer) - API Explorer extensions

Examples:

- [Minimal API](../tree/main/examples/AspNetCore/WebApi/MinimalApiExample)
- [Minimal OpenAPI](../tree/main/examples/AspNetCore/WebApi/MinimalOpenApiExample)

The discussion on route grouping is progressing well, but some help may be needed in dotnet/aspnetcore#39604. Customers don't want to wait until .NET 7 for me to release an iteration. Ideally, I'd like to release within the next month. Since things are in preview now with a lot of breaking changes from previous versions, I'm ok with a few more. I don't want to repeat that in the .NET 7 release. Anything that we can to align direction (e.g. design + signatures)  _should_ be good for all. As can be seen [here](../blob/main/src/AspNetCore/WebApi/src/Asp.Versioning.Http/Builder/ApiExplorerBuilderExtensions.cs), I've had to resort to some pretty 🤮 hackery to retain parity with the out-of-the-box _Minimal API_ support for the API Explorer.

@halter73 and I have had some discussion, but nothing is settled - yet.

**Comment by davidfowl** [comment]
Hey @commonsensesoftware, super happy to see this landing but I'm very concerned with the amount of code copied and the layering of the approach you landed on here. I think there are 2 things being coupled here that I'd like to see split up:
- The metadata for defining an API version policy, and the matcher policy
- The builder API that defines that policy for a single endpoint or group of endpoints.

What you have right now in Asp.Versioning.Http looks like its both things. This coupled with the fact that to get the desired API you've had to replicate all of the extensions (and the associated bugs) bothers me quite a bit. 

I wonder if we could have an API version policy object that could be added to multiple endpoints manually.

```C#
var builder = WebApplication.CreateBuilder(args);

// Add services to the container.
builder.Services.AddApiVersioning();

var app = builder.Build();

var apiPolicy = new ApiVersionPolicyBuilder()
            .HasApiVersion(1.0)
            .HasApiVersion(2.0)
            .ReportApiVersions()
            .Build();

// GET /weatherforecast?api-version=1.0
api.MapGet("/weatherforecast", () =>
{
        return Enumerable.Range(1, 5).Select(index =>
            new WeatherForecast
            (
                DateTime.Now.AddDays(index),
                Random.Shared.Next(-20, 55),
                summaries[Random.Shared.Next(summaries.Length)]
            ) );
})
.WithApiVersionPolicy(apiPolicy)
.MapToApiVersion(1.0);

// GET /weatherforecast?api-version=2.0
app.MapGet("/weatherforecast", () =>
{
    return Enumerable.Range(0, summaries.Length ).Select(index =>
        new WeatherForecast
        (
            DateTime.Now.AddDays( index ),
            Random.Shared.Next( -20, 55 ),
            summaries[Random.Shared.Next(summaries.Length)]
        ) );
})
.WithApiVersionPolicy(apiPolicy)
.MapToApiVersion( 2.0 );

// POST /weatherforecast?api-version=2.0
app.MapPost("/weatherforecast", (WeatherForecast forecast ) => { })
    .WithApiVersionPolicy(apiPolicy)
    .MapToApiVersion(2.0);

// DELETE /weatherforecast
app.MapDelete("/weatherforecast", () => { } )
    .WithApiVersionPolicy(apiPolicy)
    .IsApiVersionNeutral();
```

I'd prefer to have you focus on the pure API versioning data model and let us worry about how to apply policies to multiple endpoints (via route grouping).

**Comment by commonsensesoftware** [comment]
@davidfowl ,

First, thanks for taking to the time to provide feedback.

Philosophically, we are aligned. The goal of API Versioning has always been to provide ways to bolt on the metadata and let you define routes as you always have. In the most naive sense, the mental model is meant to be that API versions serve as nothing more than a way to disambiguate otherwise ambiguous endpoints. When going through the initial design review with Ryan and James, this was, and still is, achieved with a custom `IEndpointSelectorPolicy`.

Certain `404` responses as well as `405` and `415` have always been a challenge for API Versioning, but #744 really highlights the issue. I thought it might have been a mistake in the routing system itself. After some discussion with Javier, it became clear that a custom `INodeBuilderPolicy` is also necessary so that versioned endpoints are eliminated before other policies and the correct response is returned. This steams from the fact that, while you know that a candidate is invalid, you don't know _why_. The addition of a `INodeBuilderPolicy` addresses each of these problems and has an added benefit of eliminating candidates faster. This [workflow](https://viewer.diagrams.net/?tags=%7B%7D&highlight=0000ff&edit=_blank&layers=1&nav=1&title=API%20Versioning%20Flow.drawio#R3Vtbd9o4EP41PKbHVzCPgSRNT5N0z6abJo%2FCFlgbYbmyCJBfvxJIviAF3JZgZZ%2BwRrIu38x8mpFFzx%2FPV58pyNNbkkDc85xk1fMvep7nuo7Hf4RkLSWOF24lM4oSKasE9%2BgVqoZSukAJLBoNGSGYobwpjEmWwZg1ZIBSsmw2mxLcHDUHM6gJ7mOAdekPlLC0XJhTVVxDNEvl0FEoKyYgfp5RssjkeBnJ4LZmDlQ3smmRgoQsayL%2FsuePKSFs%2BzRfjSEWuCrEtu9dvVFbTpnCjLV54UcYnX8DwLl78QaT7zdfv6IvP89cf9vNC8ALiYWcLVsrcGDCsZJFQllKZiQD%2BLKSjjYAQDGOw0tVmxtCci50ufBfyNhaKh4sGOGilM2xrIUrxB7F659CWXqq1VysZM%2BbwloW9PVLSAqyoLFS7CAYeN7Qh30YDYMgOfOlNhigM8j2ICMNWCy%2B1ruE9jMkc8jomjegEAOGXpqGBKQ9zsp2lWL4g9TNG4p9a8o1Hf0Nfy5gwTRVLVPE4H0ONstfcmdtwtzUU2mQohCTOYrlMwYTiEelZY8JJpRXbWybv8YoeS7dRHQ7JRm7AnOEhWoeIE1ABqRYatzlyI0ARrOMF2KuMEj3afAFUgZXe2GXtcq51s3isubESpbW%2FFe575%2FoyexPQyv8qZVv7DP7g%2B7hRl35h3E6nu4jT3xDMUF%2FI6y7CZdmmcIAEd8ezmXFHCXJVjOwQK9gsulP6CYnKGObxYSjXnhhhH6%2FoeyaernJyVEau4XJBc6cT14YhE0%2F2JZaYy87%2F0usptaETKcFZJpyyjm00tddEi3z9OHh9tp9fvqWvz5Ov7yeuZq67sj%2F1FFcqxwl1IC%2F5mN4fczXMJpQ%2FjQTT5zFC0QyTSU0JfPJgk9qdGCrOQK3%2B01ud03k7hjIvRQeHTw%2Fas0yym7lptmhobptIx5%2F0JWlGilCj0w%2FGkW0Rt7tW8URrr6ZnmPMQ0WNJv7JihzGaIo4wBZRRRB1ThWBTrSnogqVTKlnkUpViZU5meLQ0%2FVjvVB7SxSr1zalX0rC9u6LB53D862iJe%2Fj01Jb5BUP2EJLhqARLhgF2Cbu8Z3OuUdtJzWgAjHYVeCGuqlijPICHoYKCKoXc5%2BilTDcY2AXNLHr93XsIgN00RGQM6cluokdYu13PhH7PRKvePupQdtHJ%2FHoY5K4wUO6YXHzwafTUs0n26vbqjnw7Nox9HztFrA4RdnMzpTXwpzXG9hCib9vv4p9DtNUZ4mYmab0JMImmtr%2Ffeb0u1F7NVt2eK0fdbxNU7eQa6%2FjbDu0LuQN9Q3d%2FpM5xTaHv0UGVtGSLcHTCZD3hnYxhQ79PqZIEOCNv69z2ClfhJ51fOHpkWHgWJYc%2B%2F2dU83h6bLjvc5QB822E4VwN4L2uz5SCLqgxj%2FLK5sBm%2FtuAZv64vTRbhv4ek50D7Gw5fZZZalh9zSZ5c5RW2jgYJNf9N%2BLTNRRQYOBDb7SJZkM3B3QuuXfUIfM%2FiA3aOvjYWcXJYy8re91NqXeLZnc7Z3ohLC1li07IQx0Jn87oL7Mknx7Q6rDYDqKrAumlefWQPRWK7uoPBpaFkyHv36u2j2Zq4OIw25u110iw4eAA2x%2B7MsYdgfi7fVqF32rebei7zHJCkbB5sJulww%2BcE7I4LxY%2FcVke3u3%2Bg%2BPf%2Fkf) illustrates the endpoint candidate selection process.

Building a versioning policy by hand is possible, but puts the onus on the developer and is primed for errors. Collation and construction of a policy per endpoint has always been one of the key features of API Versioning.

It's important to understand a _declared_ API version versus a _mapped_ API version. _Declaring_ an API version is what is exposed on the surface area and reported. _Mapping_ an API version only disambiguates a specific endpoint API version. Think of _mapping_ as analogous to method overriding or interface mapping; e.g. _for given endpoint `E` and method `M`, map it to API version `A`_.

Let's reconsider a sample setup with more annotations:

```c#
app.DefineApi()          // ← MapGroup (?); hope to align signatures in 6.0 so 7.0 uses final design
   .HasApiVersion( 1.0 ) // ← all endpoints in this group declare 1.0
   .HasApiVersion( 2.0 ) // ← all endpoints in this group declare 2.0
   .ReportApiVersions()  // ← all endpoints in this group reports their versions
   .HasMapping( api =>
    {
        // GET /weatherforecast?api-version=1.0
        api.MapGet( "/weatherforecast", () => new WeatherForecast[]{ new() } )
           .MapToApiVersion( 1.0 );   // ← explicitly maps to 1.0

        // GET /weatherforecast?api-version=2.0
        api.MapGet( "/weatherforecast", () => new WeatherForecast[]{ new() } )
           .MapToApiVersion( 2.0 );    // ← explicitly maps to 2.0

        // GET /weatherforecast/{id}?api-version=1.0|2.0    ← implicitly maps to 1.0 and 2.0
        api.MapGet( "/weatherforecast/{id}", ( int id ) => new WeatherForecast() );

        // POST /weatherforecast?api-version=1.0|2.0    ← implicitly maps to 1.0 and 2.0
        api.MapPost( "/weatherforecast", ( WeatherForecast forecast ) => { } );

        // DELETE /weatherforecast[?api-version=1.0|2.0]   ← implicitly maps to 1.0, 2.0, or unspecified
        api.MapDelete( "/weatherforecast", () => { } ).IsApiVersionNeutral();
    } );
```

When the `/weatherforecast` endpoint _reports its versions_, all endpoints with the exception of `DELETE` (because it's version-neutral), will report the header:

```
api-supported-versions: 1.0, 2.0
```

In MVC, this is achieved using `IActionFilter` via `ReportApiVersionsAttribute`. For Minimal APIs, it is achieved using a [decorator](../blob/main/src/AspNetCore/WebApi/src/Asp.Versioning.Http/Routing/ReportApiVersionsDecorator.cs) which is applied as a _convention_. The developer doesn't have to add each and every convention. They express their intent and the right thing happens. This invariably means customizations to the underlying builders so that a _Minimal_ API can live up to its name.

While I appreciate, and even like, the simplicity of attaching a policy, there's a bit more to it. I feel it's important that the API Versioning setup be as similar to a standard _Minimal API_ setup as possible. API Versioning has already had its own [_conventions_ API](../blob/main/examples/AspNetCore/WebApi/ConventionsExample/Program.cs) for a long time. The current design is inspired by melding those two together. I don't want the _Minimal API_ setup to be a completely new and different approach.

Ultimately, I do **not** want to copy or duplicate a bunch of intrinsic code that supports _Minimal APIs_. For the `6.0` release, I doubt there is any way around that. It's not too late for the `7.0` release though. I'm calling out the pain points and limitations I'm facing as early has possible so that there is a path to an agreeable solution. I'm highly interested and invested in following the _Grouping_ work so we are aligned. I'd also like to see the limitations in the existing design remedied. Several things are `internal`, use `internal` members, or are `sealed` for no particularly good reason. I need some help driving that.

>**Sidebar**: I faced similar challenges in the legacy `IActionSelector` design. While some of the bits were opened up in later releases, quite a bit of internal stuff had to be duplicated because there was no other way to use or consume it.

**Comment by davidfowl** [comment]
The problem is that it will exist forever and it does not compose with the existing APIs. Think about what would happen if other libraries did the same thing? They would all create their own grouping constructs that don't compose and it would be chaos (that's why we're doing it in .NET 7).

As a compromise, can you split the builder API into a separate opt-in package? 

PS: This is your package and feel free to ignore this feedback but I feel pretty strongly about this and I would argue this isn't minimal APIs, it's your own version of it. When issues show up because people move code from the top level to the version API version builder, we'll have to treat them as external bugs.

>  I'm highly interested and invested in following the Grouping work so we are aligned. I'd also like to see the limitations in the existing design remedied. Several things are internal, use internal members, or are sealed for no particularly good reason. I need some help driving that.

👍🏾

**Comment by commonsensesoftware** [comment]
@davidfowl,

Agreed. I strive to avoid forking code whenever I can. I'm not interested in owning it. I'm also aware of a few scenarios with teams who shall not be named that have not followed that recommendation and it bit them down the road. I'll just say that I'd rather not be a target of the meme:

_"When David Fowler gives you sound advice and you don't follow it."_

I'd like a mulligan. I've gone back to the whiteboard using the skeleton of what you proposed. Here's what I've come up with:

```c#
// shortcut for: app.Services.GetRequiredService<IApiVersionSetBuilderFactory>().Create(default(string));
var versionSet = app.NewApiVersionSet()
                    .HasApiVersion( 1.0 )
                    .HasApiVersion( 2.0 )
                    .ReportApiVersions()
                    .Build();

// GET /weatherforecast?api-version=1.0
app.MapGet( "/weatherforecast", () => default(WeatherForecast))
   .UseApiVersioning( versionSet )
   .MapToApiVersion( 1.0 );

// GET /weatherforecast?api-version=2.0
app.MapGet( "/weatherforecast", () => default(WeatherForecast))
   .UseApiVersioning( versionSet )
   .MapToApiVersion( 2.0 );
```

Names are important. _Policy_ sounds great, but it's a misnomer in this context. The object is not fully constructed at this point. A _policy_ would be more indicative of the final thing constructed. This _might_ be possible completely out-of-band, but it would be verbose and look ugly methinks. There also doesn't seem to be a way to reference the endpoints out-of-band to make the correlations. The API versioning information _can_ be collated across more than one endpoint.

A _version set_ seemed to more accurately represent that, but I'm not entirely married to that name. The established `WithXXX` extension methods add metadata to the endpoint. At this point in the composition, the metadata is **not** added. Only when the `Endpoint` is finally constructed is `ApiVersionMetadata` applied from all builder operations.

I hacked and slashed all the preview code. No code is forked, copied, or duplicated now. However, the existing ASP.NET Core 6.0 design has some limitations IMO.

1. The _builder_ pipeline is difficult or impossible to extend because you can only apply a convention
2. The `IEndpointConventionBuilder` has no access to services

The first issue is annoying, but it can be worked around. The second issue is more painful. Since there is neither a way to access the information that has been collected in the builder chain (say a monad of builder metadata) and there is no access to services, I was forced to bridge the two with an interface that unions the necessary behaviors. I don't see a way around this without sacrificing or changing long established features and behaviors.

```c#
public interface IVersionedEndpointConventionBuilder :
    IEndpointConventionBuilder,
    IMapToApiVersionConventionBuilder
{
    bool ReportApiVersions { get; set; }
    ApiVersionMetadata Build();
}
```

This allows the result of `UseApiVersioning` to retain the intrinsic built-in functionality, while adding API Versioning specific behaviors. This also removes builder ordering issues. `UseApiVersioning` will be required for even the most simple setup because there is no other way to get or request information such as the configured `ApiVersioningOptions`. This can't and doesn't need to change for 6.0, but at least adding `IEndpointConventionBuilder.Services` in 7.0 _should_ be considered.

Even though I don't _need_ to as a one-person show, I try to be in the habit of creating PRs so the community can see what I'm doing and potentially comment on it. As such, PR #816 is up with all the current changes I'm considering for Preview 3. A big part of the PR is this refactoring. If you'd like to comment directly on it, feel free to do so. All the changes for this topic are in a single commit. If that is TL;DR and you just want to see the changes, you can look at the source branch [here](../tree/dev/css/preview3-iteration1/src/AspNetCore/WebApi/src/Asp.Versioning.Http/net6.0/Builder). The example projects have been updated as well. Any additional thoughts or ideas are certainly welcome.

**Comment by captainsafia** [comment]
> The IEndpointConventionBuilder has no access to services

https://github.com/dotnet/aspnetcore/pull/41238 adds access to the `ServiceProvider` to the `EndpointBuilder` class which means that you can access it when adding conventions to your builder. I'm not sure if that is enough of an access point for your particular implementation but it should be available to play around with in .NET 7.

**Comment by commonsensesoftware** [comment]
@captainsafia this is great news! Having access to services, even just on the `EndpointBuilder`, will be an improvement. Armed with this knowledge, I will the necessary adjustments for the `6.0` release knowing this change will come in `7.0`. Thanks for sharing.

**Comment by commonsensesoftware** [comment]
@davidfowl, `6.0` is finally signed and published with all the of aforementioned goodness. I'll start the process of looking into adding the officially supported grouping constructs for .NET 7.0 and see how things line up.

I'll also look at the changes provided by @captainsafia for ways to streamline policy configuration and make it more natural.

Thanks for all the discussion and feedback. The project landed in a better place for it.

**Comment by davidfowl** [comment]
You are a blessing to the .NET community @commonsensesoftware! Thanks a lot!
